Pass remote debugging port for chrome

anil-groupon opened the pull request:
Set '--remote-debugging-port' in testium config, which would be consumed by lighthouse to talk with testium.

### Line of Thought: 

**Step 1**: [Indidvidual apps] Add flag `--remote-debugging-port` to _.testiumrc_ in the App's repo

**Step 2**: [testium-core] Use this flag to pass to chromedriver/chrome by either directly using the port number or by finding next available port. Also update the testium property, if new port number was used

**Step 3**: [testium-driver-wd] Add `getPort` method as wd chain methods to fetch the testium property
```
  wd.addPromiseChainMethod('getPort', function getPort() {
    let debugPort = testium.config.get('chrome.--remote-debugging-port');
    debug('Remote dubugging port set as', debugPort);
    return debugPort;
  });
```

**Step 4**: [Individual apps or itier-accessibility] Port fetch from `browser.getPort` would be used to pass it to lighthouse
```
 browser.getPort()
    .then(port => lighthouse(url, Object.assign({}, opts, { port }), config))
    .then(results => new Promise(res => res(results.audits)));
```

jkrems commented on lib/processes.js:
```diff
@@ -97,6 +97,15 @@ function launchAllProcesses(config) {
+      if (debugPort) {
+        findOpenPort().then(function (port) {
+          args.push('--remote-debugging-port=' + port);
+          config.set('chrome.--remote-debugging-port', port);
```
This feels like a weird name for an option. Why not `remoteDebuggingPort`?

jkrems commented on lib/processes.js:
```diff
@@ -97,6 +97,15 @@ function launchAllProcesses(config) {
       }
       if (process.getuid() === 0) args.push('--no-sandbox');
+
+      var debugPort = config.get('chrome.--remote-debugging-port');
```
Any reason why this isn't the default? It feels weird that people have to opt-in to enable this feature..?

anil-groupon replied:
Did you mean to always pass the port and set in the config? If yes, this was my initial thought -
In most cases, port number won't be used in the tests. Also, I was not very sure that, if we should always set the port even if it would be assigned by chromedriver.

But it makes sense now, if we are exposing API like _getPort_ from testium. We should make it return in consistent way

jkrems replied:
IIRC there was some chat to just always read it from the port file written by Chrome. Didn't that work out?

anil-groupon replied:
@jkrems Unfortunately, that didn't work. Not sure if that was due to system issue (some process errors out writing to hidden directory/files). Looking at [chromelauncher code](https://github.com/GoogleChrome/chrome-launcher/blob/master/src/chrome-launcher.ts#L235-L244), this option didn't seem bad to me :)

jkrems replied:
So setting this option is compatible with `chromedriver` after all? If so - SGTM. Is there any way we can verify that Chrome has in fact been started with that option?

anil-groupon replied:
Captured this SS from activity Monitor for the chrome instance. Port number generated programmatically was- `59196`. Not sure, if anything is obvious here :(

![screen shot 2018-07-02 at 6 14 09 pm](https://user-images.githubusercontent.com/33071225/42190717-55d1782e-7e24-11e8-94da-4e9466aa25e1.png)

jkrems commented:
Some thoughts:

1. I think `browser.getChromeDevtoolsPort()` should just work when using chrome. Having to remember to set a global option seems like it would lead to confusion in the future.
1. Implied in 1: I think `.getPort` is a little too generic for what that method does.
1. I think `testium.getChromeDevtoolsPort()` should exist and throw a helpful error message when trying to use it outside of Chrome. Accessing the config directly from `-wd` makes it harder to tell where things are going wrong.

In the future I would hope that we could even get a `driver` instance exposed from `testium` so that we can reuse more logic. AFAICT it should be possible to create a reusable `connection` (e.g. that could be used to implement better/more reliable APIs) outside of lighthouse and pass that in. So hopefully a future API would be `testium.getChromeDevtoolsConnection()` rather than `...Port()`.

anil-groupon commented:
@jkrems What is the best place to add `getChromeDevtoolsPort` method? So far, I was thinking to add in testium-driver-wd

jkrems commented:
I think the "real" implementation should definitely live in `-core`. We can expose it from `-wd` but I wouldn't want it to have deep knowledge about how it's implemented. So I'd prefer the `-wd` implementation to be not much more than `return testium.getChromeDevtoolsPort();`. That way we have all the pieces in one place and can iterate here without having to keep changes in sync.

anil-groupon commented:
@jkrems addressed the comments in last commits, except throwing error in the case port is not set. Should we just log the message for the failure case?

jkrems commented on lib/processes.js:
```diff
@@ -97,6 +103,17 @@ function launchAllProcesses(config) {
       }
       if (process.getuid() === 0) args.push('--no-sandbox');
+
+      var debugPort = config.get('chrome.remoteDebuggingPort', null);
```
Since this already runs only when we're using chrome - any reason not to default this to `true` instead of `null`?

anil-groupon replied:
@jkrems This assumes that if the port has been set (with actual value) in _.testiumrc_, it would address that. Otherwise would discover the new port and update the config.

Are you suggesting that to remove the check and find new port anyway?

jkrems replied:
Is there value in making the port configurable? That seems to encourage hard-coding values which might lead to problems when running multiple test suites etc..

anil-groupon replied:
Oh yes. I didn't think about that. I'll remove the checks

jkrems commented on lib/init.js:
```diff
@@ -127,11 +127,16 @@ function initTestium(config) {
       getInitialUrl: getInitialUrl,
-      getNewPageUrl: _.partial(getNewPageUrl, config.get('proxy.targetUrl'))
+      getNewPageUrl: _.partial(getNewPageUrl, config.get('proxy.targetUrl')),
+      getChromeDevtoolsPort: getChromeDevtoolsPort
```
nit: it looks like those were sorted alphabetically before.

jkrems commented:
Thanks for digging into this & sorry for the long back-and-forth! :)

anil-groupon commented:
@jkrems Thanks for the review :)

How do I release new version now? I believe one with 'write access' can do

jkrems commented:
These libraries are under npm, so the new version got released once this got merged!
